### PR TITLE
Haah's cubic code using abelian group ℤ₃ˣ³ via Hecke's Group Algebra

### PR DESCRIPTION
**Pull request by Fe-r-oz**
This PR aims to implement Haah's cubic code, a cousin of Generalized Bicycle code as mentioned in https://errorcorrectionzoo.org/c/generalized_bicycle.

- [x] The code is properly formatted and commented.
- [x] Substantial new functionality is documented within the docs.
- [x] All new functionality is tested.
- [x] All of the automated tests on github pass.
- [x] We recently started enforcing formatting checks. If formatting issues are reported in the new code you have written, please correct them. <small>There will be plenty of old code that is flagged as we are slowly transitioning to enforced formatting. Please do not worry about or address older formatting issues -- keep your PR just focused on your planned contribution.</small>

### CODE REVIEW

**Comment by Fe-r-oz** [comment]
I think this is ready for review, @krastanov. Thanks!

Brief context: For GB codes , `n%l+1` is used for GroupAlgebra (GA) indexing to indicate the cyclic nature for single cyclic group. This is same as doing  `n%dim(GA)+1.` This works as here, we have a single cyclic group of order l, so total order is l which can be acquired using `dim(GA)`.  

In Haah cubic code case, the total order is `l³`, hence using `n%dim(GA)+1` is appropriate since `dim(GA)` would give `l³` because here, direct product of three cyclic groups ( `total order = lxlxl`) is the requirement. Also, The CI errors are unrelated to this PR.

File: ext/QuantumCliffordHeckeExt/lifted_product.jl [hunk 1]
```diff
@@ -216,3 +216,23 @@ function bicycle_codes(a_shifts::Array{Int}, l::Int)
     a = sum(GA[n÷l+1] for n in a_shifts)
     two_block_group_algebra_codes(a, group_algebra_conj(a))
 end
+
+"""
+Haah’s cubic codes [haah2011local](@cite) can be viewed as GB codes
```
Comment (Krastanov) [inline ext/QuantumCliffordHeckeExt/lifted_product.jl:221]:
```suggestion
Haah’s cubic codes [haah2011local](@cite) can be viewed as generalized bicycle (GB) codes
```

File: ext/QuantumCliffordHeckeExt/lifted_product.jl [hunk 2]
```diff
@@ -216,3 +216,23 @@ function bicycle_codes(a_shifts::Array{Int}, l::Int)
...
+with the group `G = Cₗ × Cₗ × Cₗ`, where `l` denotes the lattice size.
+In particular, a GB code with the group `G = ℤ₃ˣ³` corresponds to a
+cubic code [https://errorcorrectionzoo.org/c/haah_cubic](@cite).
+
+```jldoctest
+julia> c = Haah_cubic_codes([0, 15, 20, 28, 66], [0, 58, 59, 100, 121], 6);
+
+julia> code_n(c), code_k(c)
+(432, 8)
+```
```
Comment (Krastanov) [inline ext/QuantumCliffordHeckeExt/lifted_product.jl:231]: see also: bicycle_codes, generalized_bicycle_codes

File: ext/QuantumCliffordHeckeExt/lifted_product.jl [hunk 3]
```diff
@@ -216,3 +216,23 @@ function bicycle_codes(a_shifts::Array{Int}, l::Int)
...
+"""
+function Haah_cubic_codes(a_shifts::Array{Int}, b_shifts::Array{Int}, l::Int)
```
Comment (Krastanov) [inline ext/QuantumCliffordHeckeExt/lifted_product.jl:233]:
```suggestion
function haah_cubic_codes(a_shifts::Array{Int}, b_shifts::Array{Int}, l::Int)
```

a bit on the fence about this change, but it seems more consistent with other naming conventions in the library

**Review by Krastanov — CHANGES_REQUESTED** [review]
Another great addition, thank you! I left in a few comments.

**Comment by Fe-r-oz** [comment]
The CI is failing on master due to this: 

https://github.com/QuantumSavory/QuantumClifford.jl/blob/45e9f4d64c902ba95ae5bedc1e59895f552147c4/ext/QuantumCliffordHeckeExt/lifted_product.jl#L253-L256

This was fixed with the the see also.

**Comment by Krastanov** [comment]
> The CI is failing on master due to this:
> 
> https://github.com/QuantumSavory/QuantumClifford.jl/blob/45e9f4d64c902ba95ae5bedc1e59895f552147c4/ext/QuantumCliffordHeckeExt/lifted_product.jl#L253-L256
> 
> This was fixed with the the see also.
> 
> Edit: It seems that theCI is not failing due to the extra spaces, I thought it would cause error when checking docstring.

that fix is still nice to have!

I think it is failing because there are still a few spots with the old capitalization of the name of the code

**Comment by Krastanov** [comment]
marking as draft to make it easier for me to track what needs review

**Comment by Fe-r-oz** [comment]
The PR is ready for review. Thank you! 

It was indeed the old capitalization in some places that was causing the CI  errors.

File: ext/QuantumCliffordHeckeExt/lifted_product.jl [hunk 4]
```diff
@@ -302,10 +301,33 @@ Bicycle codes are a special case of generalized bicycle codes,
...
+
+"""
+Haah’s cubic codes [haah2011local](@cite) can be viewed as generalized bicycle (GB) codes
+with the group `G = Cₗ × Cₗ × Cₗ`, where `l` denotes the lattice size. In particular, a GB
+code with the group `G = ℤ₃ˣ³` corresponds to a cubic code.
+
+The ECC Zoo has an [entry for this family](https://errorcorrectionzoo.org/c/haah_cubic).
+
+```jldoctest
+julia> c = Haah_cubic_codes([0, 15, 20, 28, 66], [0, 58, 59, 100, 121], 6);
```
Comment (Krastanov) [inline ext/QuantumCliffordHeckeExt/lifted_product.jl:320]:
```suggestion
julia> c = haah_cubic_codes([0, 15, 20, 28, 66], [0, 58, 59, 100, 121], 6);
```

**Pull request merged**